WIP: Ensure `rails new . --master` builds a complete bleeding edge application (not just core gems)

Schwad opened the pull request:
### Summary

In PR #38631 I implemented support for the `--master` option in `rails new`. This allows quick and easy functionality for devs to test bleeding edge Rails features as soon as they are committed. (Prior to that we had `--edge` which only pointed to latest stable. If ran from within the master Rails version it would point to master branch, but not otherwise)

There was an ensuing discussion with @matthewd where he pointed out that in the future this could become quite a sharp tool. Say in 2022 someone ran the `--master` flag from a `6.0` rails version but master was on `6.2` which had a radically different file structure. You could have a situation where you have a file structured like `6.0` but pointing to `6.2` Gemfile.

@rafaelfranca  currently has retained the feature for the next release, but we have had a discussion about improving this flag to make it more unique than `--edge` and `--dev`. 

The pitched idea was to have the flag manually write a bleeding edge Gemfile pointing to master, `bundle install` it, and then use that to grab the latest rails with `bundle exec`, but all invoked with the current interface of `rails new . --master`.

This pull request is a WIP attempt to continue this discussion and pull this idea together. It currently doesn't work, but that's likely more due to my inexperience with building out novel new generators within rails than the quality of the idea.

It is a small PR, and it should be clear what I'm trying to accomplish. I would love feedback especially from the prior discussion participants (including @kaspth ).

I'm also available to pair on this (I have a Tuple license I can invite from) sometime this week or next week.

Thanks much again for the quality discussion around this! ❤️

Schwad committed:
Allow complete app generation off of master rails

Schwad committed:
Some puts debugging

Schwad committed:
Merge branch 'master' of https://github.com/Schwad/rails

Schwad committed:
Try to pass in as a builder class separately

Schwad committed:
Rename

Schwad committed:
Rename

Schwad committed:
Rename

Schwad committed:
One last punt

Schwad committed:
Cleanup puts debugging for external review

Schwad commented:
Note: Obviously this would also require me to revert the prio `--master` implementation (which would be redundant with this). I will include that commit if we're able to come to a working consensus here.

Schwad committed:
Reorganize and rename code with proper hierarchy

Schwad committed:
Debugging test script

Schwad committed:
Debug two

Schwad committed:
Lazy retry debug

Schwad committed:
Bump

Schwad committed:
Stop debugging and show as example to OSS folks

kaspth commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -243,6 +249,17 @@ def config_target_version
 
+  class MasterGemfileGenerator < AppBuilder # :nodoc:
+    def install_shell_gemfile
+      template "GemfileMasterRailsInstaller", "Gemfile"
```
So we're using this to essentially build a local Rails checkout that can then generate an application?

Not really liking installing this Gemfile that'll linger afterward. We may as well just do a shallow clone via `git clone https://github.com/rails/rails.git`. We essentially don't need any other lines in the file and if we install the checkout at a known location, we can do `/path/to/rails/railties/exe/rails new …`

kaspth commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -243,6 +249,17 @@ def config_target_version
+    end
+
+    def generate_edge_rails_app
+      run 'bundle exec rails new . --edge'
```
Can you actually pass `.` and it'll generate into the current directory without naming the app "."? Think we also want to pass on extra options ala `--webpack=stimulus`.

kaspth commented on railties/lib/rails/generators/rails/app/templates/GemfileMasterRailsInstaller.tt:
```diff
@@ -0,0 +1,6 @@
+source 'https://rubygems.org'
```
Not liking the camelcase naming on the file I gotta say 😄

kaspth commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -243,6 +249,17 @@ def config_target_version
     end
   end
 
+  class MasterGemfileGenerator < AppBuilder # :nodoc:
```
Is the `Generator` suffix right for this?

kaspth commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -7,11 +7,17 @@ module ActionMethods # :nodoc:
     end
 
     private
+
```
Remove this.

kaspth commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -7,11 +7,17 @@ module ActionMethods # :nodoc:
-      @options   = generator.options
+      if generator.options[:master]
+        puts '== Generating new rails app based off of rails/rails master branch =='
+        MasterGemfileGenerator.new
```
Hm, not exactly liking this call structure, but I don't know where else to put this.

kaspth commented:
I'm wondering if we don't want a one file shell script that we can run instead of hooking in multiple places in our generators setup? E.g.

```sh
#!/usr/bin sh
set -ex

git clone https://github.com/rails/rails ~/.rails/rails # Or where ever's better. This'll update the clone if it exists, right?

app_name=shift
~/.rails/rails/railties/exe/rails new $app_name --edge $@ # Or however the hell you pass the remaining arguments
```

Commit:
Rename gemfile master rails installer to snake case

Commit:
Include more appropriate name for MasterRails generator

Commit:
Style error fix

Commit:
Renaming file and options comment

Commit:
Look into simple script

Schwad commented:
@kaspth thank you for taking a look and your thoughts! 😅

My first few commits addressed some of your feedback, and the last one is a slight peek at possibly going down the shell script route. I actually really like that idea. Unless you change your mind I might nuke my old approach and proceed thusly.

Obviously 'testing' this manually is... interesting.. but at least our friends gave us some test coverage of this command so as long as we pass that we should be in good shape. 😇
